boot/bootstate20.go

```diff
@@ -21,330 +21,526 @@ package boot
+		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
-	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
-	if !ok {
-		return nil, nil, false, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", ebl.Name())
+	trying = (m["kernel_status"] == "trying")
+
+	return bootSn, tryBootSn, trying, nil
+}
+
+func (ks20 *bootState20Kernel) setNext(next snap.PlaceInfo) (bool, bootStateUpdate, error) {
+	err := ks20.loadBootloader()
+	if err != nil {
+		return false, nil, err
 	}
 
-	// get the kernel for this bootloader
-	bootSnap, err = ebl.Kernel()
+	r, err := genericSetNext(ks20, next)
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ebl.Name(), err)
+		return false, nil, err
 	}
 
-	tryKernel, tryKernelExists, err := ebl.TryKernel()
+	return r, ks20, nil
+}
+
+func (ks20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	// call the generic method with to do most of the legwork
+	u, sn, err := genericMarkSuccessful(
+		ks20,
+		update,
+	)
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ebl.Name(), err)
+		return nil, err
 	}
 
-	if tryKernelExists {
-		tryBootSnap = tryKernel
+	// u should always be non-nil if err is nil
+	u.triedKernelSnap = sn
+	return u, nil
+}
+
+// commit for bootState20Kernel is meant only to be used with setNext(), for
+// markSuccessful(), use bootState20MarkSuccessful.
+func (ks20 *bootState20Kernel) commit() error {
+	// The ordering of this is very important for boot safety/reliability!!!
+
+	// If we are about to try an update, and need to add the try-kernel symlink,
+	// we need to do things in this order:
+	// 1. Add try-kernel symlink
+	// 2. Update kernel_status to "try"
+	//
+	// This is because if we get rebooted in between 1 and 2, kernel_status
+	// is still unset and boot scripts proceeds to boot with the old kernel,
+	// effectively ignoring the try-kernel symlink.
+	// If we did it in the opposite order however, we would set kernel_status to
+	// "try" and then get rebooted before we could create the try-kernel
+	// symlink, so the bootloader would try to boot from the non-existent
+	// try-kernel symlink and become broken.
+
+	// add the try-kernel symlink
+	// trySnap could be nil here if we called setNext on the current kernel
+	// snap
+	if ks20.tryKernelSnap != nil {
+		err := ks20.ebl.EnableTryKernel(ks20.tryKernelSnap)
+		if err != nil {
+			return err
+		}
 	}
 
-	m, err := bl.GetBootVars("kernel_status")
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ebl.Name(), err)
+	m := map[string]string{
+		"kernel_status": ks20.kernelStatus,
 	}
 
-	trying = m["kernel_status"] == "trying"
+	// set the boot variables
+	return ks20.ebl.SetBootVars(m)
```

Comment:
did this code lose the property of not writing the bootenv if nothing needs changing? maybe it was not tested but we do want that property

the old code had a check like: `if len(u20.toCommit) == 0`

Reply:
hmm I suppose this has changed, I'm assuming that's a desirable property, so I will refactor this a bit to not write if the bootenv doesn't need changing.

Reply:
I implemented this and added an additional check for this to one of the tests.